### PR TITLE
New cop Style/SafeNavigation

### PR DESCRIPTION
**Pull request by rrosenblum**
This addresses #2428.

I am just about done with this new cop. Functionally, I think that I am set. I have run into a name spacing issue though. The name spacing issue is the root of the test failures. `qualified_cop_name` does not like that there are 2 cops with the same base name (`Style/SafeNavigation` and `Rails/SafeNavigation`). I am not sure how we would like to handle this. 

The solutions that I have considered are:

1. Change the name of either the `Style/SafeNavigation` or `Rails/SafeNavigation` cop. 

I am unsure of a good name to change either of these to. If anyone has ideas, please let me know.

2. Modify the `qualified_cop_name` method to do a lookup using the cop type as well as the name of the cop. 

This would fix occurrences that use the full cop name including the group. It would not fix occurrences where only the cop name is given. For example, how would we handle `rubocop --only SafeNavigation`? Should that run both the `Rails` and `Style` cop? This would also impact using the comment disable of a cop.

### CODE REVIEW

File: spec/rubocop/cop/style/safe_navigation_spec.rb [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+            expect(new_source).to eq('puts 1 if foo&.bar(baz) { |e| e.baz }')
+          end
+        end
+      end
+    end
+  end
+
+  context 'target_ruby_version < 2.3', :ruby19 do
+    it 'registers an offense for method calls safe guarded by a ' \
```
Comment (daviddavis) [inline spec/rubocop/cop/style/safe_navigation_spec.rb]: I think this should be "it does not register an offense ..." or "it registers no offense ..."

File: lib/rubocop/cop/style/safe_navigation.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# encoding: utf-8
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop transforms usages of a method call safe guarded by a call to
+      # `respond_to?` to safe navigation `&.`.
```
Comment (bbatsov) [inline lib/rubocop/cop/style/safe_navigation.rb]: You should add here some examples of code that would be flagged by this cop.

File: config/enabled.yml [hunk 3]
```diff
@@ -664,6 +664,12 @@ Style/RescueModifier:
   StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#no-rescue-modifiers'
   Enabled: true
 
+Style/SafeNavigation:
+  Description: >-
+                  This cop transforms usages of a method call safe guarded by
```
Comment (bbatsov) [inline config/enabled.yml]: safeguarded

**Comment by bbatsov** [comment]
> I am unsure of a good name to change either of these to. If anyone has ideas, please let me know.
> 
> Modify the qualified_cop_name method to do a lookup using the cop type as well as the name of the cop.
> This would fix occurrences that use the full cop name including the group. It would not fix occurrences where only the cop name is given. For example, how would we handle rubocop --only SafeNavigation? Should that run both the Rails and Style cop? This would also impact using the comment disable of a cop.

I'd simply raise some error in such cases where some ambiguity exists.

File: spec/rubocop/cop/style/safe_navigation_spec.rb (hunk 1 above)
Comment (rrosenblum) [inline spec/rubocop/cop/style/safe_navigation_spec.rb]: It appears that `accepts` is preferred over `does not` throughout the test suite.

```
rubocop (safe_navigation u=)$ grep "accepts" spec -r | wc -l            
    1332
rubocop (safe_navigation u=)$ grep 'does not' spec -r | wc -l
     539
rubocop (safe_navigation u=)$ grep "doesn't" spec -r | wc -l
     214
```

File: lib/rubocop/cop/style/safe_navigation.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop transforms usages of a method call safeguarded by a call to
+      # `respond_to?` to safe navigation `&.`.
+      #
+      # @example
+      #   # bad
+      #   foo.bar if foo.respond_to?(:bar)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/safe_navigation.rb]: Shouldn't this be `foo.bar if foo`? We care only about redundant nil checks here after all. 
`foo && foo.bar` is another classic pattern. As is `foo.bar unless foo.nil?`.

Comment (rrosenblum) [inline lib/rubocop/cop/style/safe_navigation.rb]: I think that you are correct. I think I mixed up the functionality of for safe navigation as it pertains to a `nil` check vs a `respond_to?` check. I will need to update all of the functionality of this cop. This change should not be difficult to make. Good catch.

**Comment by rrosenblum** [comment]
I finally finished up this cop after having coded it incorrectly the first time.

The cop will check for and correct code in the following formats:
```ruby
foo.bar if foo 
foo.bar if !foo.nil?
foo.bar unless !foo
foo.bar unless foo.nil?

foo && foo.bar
!foo.nil? &&
foo.nil? || foo.bar
!foo || foo.bar
```

As with the `Rails/SafeNavigation` cop, this cop will not register an offense if the method be called is a method that `nil` responds to. It also will not register an offense for a method that looks weird when combined with `&.`.

The cop name qualifier now checks for an exact match to the cop, including the group, before checking in the existing manner with the cop name only. This is so that we can have 2 cops with the same name in 2 different groups. If a cop name is found to be ambiguous, the names of the matching cops will be printed out in the error message. This also means that we finally get to enable the test for ambiguous cop names.

**Comment by bbatsov** [comment]
Great work! Rebase this and let's have it merged.

**Comment by bbatsov** [comment]
The build is failing.

**Comment by rrosenblum** [comment]
The build is now passing.

**Pull request merged**